text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core` libs changes
+
+The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.
+
+This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.
+
+```rust
+pub trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+A blanket implementation is provided for any type that implements `Deref`:
+
+```rust
+impl<P: ?Sized, T: ?Sized> Receiver for P
+where
+    P: Deref<Target = T>,
+{
+    type Target = T;
+}
+```
+
+(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)
+
+It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.
+
+## Compiler changes
+
+The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.
+
+Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.
+
+Dereferencing a raw pointer usually needs `unsafe` (for good reason!) but in this case, no actual dereferencing occurs. This is used only to determine a list of method candidates; no memory access is performed and thus no `unsafe` is needed.
+
+## Object safety
+
+Receivers are object safe if they implement the (unstable) `core::ops::DispatchFromDyn` trait.
+
+As not all receivers might want to permit object safety or are unable to support it, object safety should remain being encoded in a different trait than the here proposed `Receiver` trait, likely `DispatchFromDyn`.
+
+This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.
+
+## Lifetime elision
+
+As discussed in the [motivation](#motivation), this new facility is _most likely_ to be used in cases where a standard reference can't normally be used. But in other cases a smart pointer self type might wrap a standard Rust reference, and thus might be parameterized by a lifetime.
+
+Lifetime elision works in the expected fashion:
```

Comment:
How does it work if the pointer has more than one lifetime?

Reply:
I can't spot any problems here - [here's a playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=867bae8b8d1e9e97680f02e972a59fd7). This is of course with the current unstable `arbitrary_self_types` but nothing in this area should change. If you can think I should clarify something in the RFC text please let me know.

Reply:
What I meant is: how does lifetime elision work then? There's no "expected fashion" for this case since it's not currently possible. So when I write `fn foo<'a, 'b, 'c>(self: MyPtr<'a, 'b, Self>, other: &'c T) -> &U`, then which lifetime does the returned reference get? It's ambiguous. I think this should be rejected and require an explicit lifetime.

Reply:
I would imagine that the lifetime elision rules would be augmented this way:

If the self type has exactly one lifetime, then the rules behave the same as `&self`. So, this:
```rust
fn method(self: P<'_, Self>, one: &T, two: &T) -> &T
```
Is sugar for:
```rust
fn method<'selflt, 'a, 'b>(self: P<'selflt, Self>, one: &'a T, two: &'b T) -> &'selflt T
```

If the return type has no lifetime, then the inference assumes that all elided lifetimes are distinct, even if `self` has multiple lifetimes. So, this:

```rust
fn method(self: P<'_, '_, Self>, one: &T, two: &T) -> Box<T>
```
Is sugar for:
```rust
fn method<'a, 'b, 'c, 'd>(self: P<'a, 'b, Self>, one: &'c T, two: &'d T) -> Box<T>
```

If the self type has multiple lifetimes, then it's flagged as ambiguous if the return type has any lifetimes. So, this is ambiguous:
```rust
fn method(self: P<'_, '_, Self>, one: &T, two: &T) -> &T
```

Reply:
> If the self type has multiple lifetimes, then it's flagged as ambiguous if the return type has any lifetimes. So, this is ambiguous:

Yeah sounds good. The same should probably occur if `self` has no lifetime (like it already does for `self: Self`, I assume).

Reply:
I rewrote this section to include an example much like that playground link. Please take another look!

Reply:
> However, I still believe the existing lifetime elision rules work OK with no changes. [Here's a playground using the existing unstable arbitrary_self_types support](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=afe43820163195ff26454e74f3c2881c) - if there's any ambiguity, the compiler insists that a lifetime is manually specified.

Those are not the "existing rules". The existing rules are what is documented in the reference or prior RFCs.

You are saying the compiler already implements the rules you are suggesting, which is great, but that doesn't mean those rules have been subject to any kind of scrutiny, and unless you can point at existing documentation detailing those rules, it is the job of this RFC to document them.